docs/CONFIGURATION_DOC.md
# Configuration Documentation


## Config structure
```json5
{
  "CONFIG_VERSION_DONT_TOUCH_THIS": 1,
  
  "general": {
    
    // Teleports the player to the plot world automatically on respawn
    "autoTeleport": true, // boolean
    
    // Displays a warning in case the player sends a message in the plot chat and no one else is on the plot
    "warningUseOfPlotChatInEmptyPlot": true, // boolean
    
    // Show plot chat messages in the logs
    "logPlotChat": true, // boolean

    // Coordinates that are added to the spawn (where you go with /plot visit) of each plot 
    // if the spawn has not changed (with /plot setspawn)
    "defaultPlotSpawn": { // Block Pos
      "x": 0, // int
      "y": 0, // int
      "z": 0 // int
    },
    
    // List of entities that are allowed to spawn, any other entity in the world will be removed
    "entityWhitelist": [ // Entity ID
      "armor_stand"
    ],
    
    // List of entities that are allowed to spawn on a road or to leave a plot
    "entityRoadWhitelist": [ // Entity ID
      "item"
    ]
  },

  // Settings related to world generation, it is not recommended to modify them once you have the server in production 
  // as it can cause inconsistencies
  "PlotWorld": {
    
    // Distance between the borders of the plots, border is not counted
    "roadSize": 8, // int
    
    // Plot size between borders, border is not counted
    "plotSize": 200, // int
    
    // ID of the biome to use, if not found plains is used
    // It only changes when the server is restarted
    "biomeId": "minecraft:plains",
    
    // The block that is one higher up between the road and the plot when the plot does not have an owner
    "unclaimedBorderBlock": { // Block State
      "id": "stone_slab",
      "states": []
    },
    
    // The block that is one higher up between the road and the plot when the plot has an owner
    "claimedBorderBlock": { // Block State
      "id": "deepslate_tile_slab",
      "states": []
    },
    
    // The block that is between the road and the plot
    "borderBlock": { // Block State
      "id": "deepslate_tiles",
      "states": []
    },
    
    // The block used on roads
    "roadBlock": { // Block State
      "id": "white_concrete",
      "states": []
    },
    
    // Layers of blocks that are in the plots, ordered from bottom to top
    "layers": [
      {
        
        // The number of layers of this block
        "thickness": 1,
        
        "block": { // Block State
          "id": "barrier",
          "states": []
        }
      }
    ]
  },
  
  "database": {
    
    // Database type
    "type": "h2", // h2 (soon more options)
    
    // Username to enter the database
    "user": "root", // string
    
    // Password to enter the database user
    // Leave the field empty if a password is not required
    "password": "", // string
    
    // The name of the database
    "database": "plotcubic", // string
  },

  // Chat style affects "/plot chat" for everyone within a plot
  "plotChatStyles": [
    {
      
      // The ID is used to differentiate one chat style from another.
      // You can give a user permission to use a chat style with plotcubic.chatstyle.style.ID
      "id": "frozen_berries", // string preferably with only lowercase letters and underscores
      
      // Chat style name
      "name": "Frozen Berries", // string
      
      // ID of the chat style item that the user sees in the GUI
      "itemId": "sweet_berries", // Minecraft item id
      
      // Message displayed when using plot chat with this chat style enabled
      // Support PlaceHolder API
      // Placeholders:
      // To show the ID of the plot: %plot_id% for example: 3;50
      // To show the user who sent the message: %username%
      // To display the message: %message%
      "message": "<color:#595e60>[<color:#707370>%plot_id%<color:#595e60>] <color:#b0b7c0>%username% <color:#707370>» <color:#b0b7c0>%message%" // string
    }
  ],
  
  // Colors used by PlotCubic
  "customColors": {
    
    // Color of the parameters in the messages
    "highlight": "2C8395", // hex color
    
    // They are used to avoid having hardcode colors in translations.
    "others": [
      {
        
        // Color name, used with <NAME> in translation files
        "name": "p3_normal",  // string preferably with only lowercase letters and underscores
        
        // Color value as hexadecimal number
        "color": "61C2A2" // hex color
      }
    ]
  },
  
  // Possible reporting reasons
  "reportReasons": [
    {
      // Report Reason ID
      // The translation of the name goes in text.plotcubic.report_reason.ID.name
      // The translation of the descriptions goes in text.plotcubic.report_reason.ID.description.INDEX
      "id": "griffed", // string preferably with only lowercase letters and underscores
 
      // Report descriptions go below the name
      // The report descriptions go below the name, this value indicates the number of lines, starts at 0 and increases
      // 0 if there is no description
      "descriptionCount": 0, // int
      
      // The item used in the GUI to report
      "item": { // GUI Item 
        "itemId": "lava_bucket",
        "count": 1,
        "hideAttributes": false,
        "glow": false,
        "headValue": ""
      }
    }
  ]
}
```

## Reload

You can reload the configuration without restarting the server with `/plot admin reload`

## Objects structure

### Block state
You can find more information in https://minecraft.fandom.com/wiki/Block_states

```json5
{
  // Block ID, can contain namespace
  "id": "stone_slab",
  
  // Block states are properties that usually change the texture and/or hitbox of the block.
  // You can see the block state of a block by looking at it in F3, in Targeted Block on the right,
  // they are the first after the block ID and are divided into <key>: <value>
  // Note: not all blocks have states, and if they are not specified, the default ones are used.
  "states": [
    {
      "key": "waterlogged",
      "value": "true"
    }
  ]
}
```

### GUI Item


```json5
{
    
    "itemId": "lava_bucket", // item ID
    "count": 1, // int
    
    // Hide attributes of an item, it is recommended to activate in tools and armor
    "hideAttributes": false, // boolean
    
    // Activate enchantment glow
    "glow": false, // boolean
    
    // The Value tag contains the skin of a head, it is found in SkullOwner.Properties.textures[0].Value
    // Or you can upload a skin to get the value at https://mineskin.org
    // You can leave the value empty in case it is not a head
    "headValue": "" // string
}
```

src/main/java/me/zailer/plotcubic/config/Config.java
package me.zailer.plotcubic.config;

import me.zailer.plotcubic.plot.PlotChatStyle;
import me.zailer.plotcubic.plot.ReportReason;

public record Config(int CONFIG_VERSION_DONT_TOUCH_THIS, General general, PlotworldConfig PlotWorld,
                     Database database, PlotChatStyle[] plotChatStyles, CustomColors customColors,
                     ReportReason[] reportReasons) {

    public static final Config DEFAULT = new Config(
            1,
            new General(true,
                    true,
                    true,
                    new BlockPos(0, 0, 0),
                    new String[] { "armor_stand", "arrow", "boat", "chest_minecart", "egg", "end_crystal", "ender_pearl", "experience_bottle", "eye_of_ender", "firework_rocket", "fishing_bobber", "furnace_minecart", "glow_item_frame", "item", "item_frame", "minecart", "painting", "potion", "snowball", "spectral_arrow", "tnt", "trident" },
                    new String[] { "eye_of_ender", "item" }
            ),
            new PlotworldConfig(
                    8,
                    200,
                    "minecraft:plains",
                    new BlockConfig("stone_slab", new BlockStateConfig[]{}),
                    new BlockConfig("deepslate_tile_slab", new BlockStateConfig[]{}),
                    new BlockConfig("deepslate_tiles", new BlockStateConfig[]{}),
                    new BlockConfig("white_concrete", new BlockStateConfig[]{}),
                    new Layer[]{
                            new Layer(1, new BlockConfig("barrier", new BlockStateConfig[]{})),
                            new Layer(49, new BlockConfig("stone", new BlockStateConfig[]{})),
                            new Layer(1, new BlockConfig("grass_block", new BlockStateConfig[]{}))
                    }
            ),
            new Database(
                    "h2",
                    "root",
                    "123",
                    "plotcubic"
            ),
            new PlotChatStyle[]{
                    new PlotChatStyle("frozen_berries", "Frozen Berries", "sweet_berries",
                            "<color:#595e60>[<color:#707370>%plot_id%<color:#595e60>] <color:#b0b7c0>%username% <color:#707370>» <color:#b0b7c0>%message%"),
                    new PlotChatStyle("window_tile", "Window Tide", "cyan_stained_glass_pane",
                            "<color:#41729f>[<color:#5885af>%plot_id%<color:#41729f>] <color:#c3e0e5>%username% <color:#5885af>» <color:#c3e0e5>%message%"),
                    new PlotChatStyle("rosy_flamingo", "Rosy Flamingo", "pink_dye",
                            "<color:#67595e>[<color:#a49393>%plot_id%<color:#67595e>] <color:#e8b4b8>%username% <color:#eed6d3>» <color:#e8b4b8>%message%"),
                    new PlotChatStyle("healthy_leaves", "Healthy Leaves", "azalea_leaves",
                            "<color:#3d550c>[<color:#81b622>%plot_id%<color:#3d550c>] <color:#ecf87f>%username% <color:#81b622>» <color:#ecf87f>%message%"),
                    new PlotChatStyle("candy_hearts", "Candy Hearts", "cake",
                            "<color:#fceca5>[<color:#fec8a7>%plot_id%<color:#fceca5>] <color:#b5eecb>%username% <color:#cbc7fc>» <color:#b5eecb>%message%"),
                    new PlotChatStyle("sweeter_love", "Sweeter Love", "red_dye",
                            "<color:#838bc2>[<color:#a58cb3>%plot_id%<color:#838bc2>] <color:#d9a1a0>%username% <color:#eae7fa>» <color:#d9a1a0>%message%"),
                    new PlotChatStyle("halloween", "Halloween", "jack_o_lantern",
                            "<color:#ccc5b9>☠<color:#fffcf2>%plot_id%<color:#ccc5b9>☠ <color:#eb5e28>%username% <color:#eca72c>⚡ <color:#eb5e28>%message%"),
                    new PlotChatStyle("christmas", "Christmas", "spruce_sapling",
                            "<color:#6a994e>⛄<color:#a7c957>%plot_id%<color:#6a994e>⛄ <color:#f2e8cf>%username% <color:#bc4749>★ <color:#f2e8cf>%message%"),
                    new PlotChatStyle("valentines_day", "Valentine's Day", "bow",
                            "<color:#ed4545>❤<color:#fbc6cc>%plot_id%<color:#ed4545>❤ <color:#cab5c8>%username% <color:#ed4545>\uD83C\uDFF9 <color:#cab5c8>%message%"),
                    new PlotChatStyle("orange_tulip", "Orange tulip", "orange_tulip",
                            "<color:#264653>[<color:#2a9d8f>%plot_id%<color:#264653>] <color:#e9c46a>%username% <color:#f4a261>❁ <color:#e76f51>%message%"),
                    new PlotChatStyle("fire", "Fire", "netherrack",
                            "<color:#f1e8b8>[<color:#f9e784>%plot_id%<color:#f1e8b8>] <color:#e58f65>%username% <color:#d05353>\uD83D\uDD25 <color:#e58f65>%message%")
            },
            new CustomColors(
                    "2C8395",
                    new Color[]{
                            new Color("p3_normal", "61C2A2"),
                            new Color("p3_icon", "CCE0D2"),
                            new Color("p3_error", "1D617A"),
                            new Color("p3_red", "A06767"),
                            new Color("p3_green", "67A067"),
                            new Color("p3_blue", "6767A0")
                    }
            ),
            new ReportReason[]{
                    new ReportReason("griffed", 0, new ItemConfig("lava_bucket", 1, false, false, "")),
                    new ReportReason("explicit_sexual_content", 0, new ItemConfig("barrier", 18, false, false, "")),
                    new ReportReason("ip_spam", 0, new ItemConfig("oak_sign", 1, false, false, "")),
                    new ReportReason("generates_lag", 0, new ItemConfig("redstone", 1, false, false, "")),
                    new ReportReason("hate_speech", 0, new ItemConfig("player_head", 1, false, false, "eyJ0ZXh0dXJlcyI6eyJTS0lOIjp7InVybCI6Imh0dHA6Ly90ZXh0dXJlcy5taW5lY3JhZnQubmV0L3RleHR1cmUvYzAwMThhNDNjYzQ0OGRmZGU5ZWU0Mjg4ZDZjZTQyMWM5NmU4MTQyZmY4YzE5NWM0NDRlMGUxMTg0ZWNmY2M1NSJ9fX0=")),
                    new ReportReason("spawn_kill", 0, new ItemConfig("iron_sword", 1, true, false, "")),
                    new ReportReason("other", 0, new ItemConfig("paper", 1, false, false, ""))
            }
    );

    public record General(boolean autoTeleport, boolean warningUseOfPlotChatInEmptyPlot,
                          boolean logPlotChat, BlockPos defaultPlotSpawn, String[] entityWhitelist,
                          String[] entityRoadWhitelist) {
    }

    public record Database(String type, String user, String password, String database) {
    }

    public static final class CustomColors {
        private String highlight;
        private final Color[] others;

        public CustomColors(String highlight, Color[] others) {
            this.highlight = highlight;
            this.others = others;
        }

        public void setDefaultHighlight() {
            this.highlight = "2C8395";
        }

        public String highlight() {
            return highlight;
        }

        public Color[] others() {
            return others;
        }
    }

    public record Color(String name, String color) {
    }

    public record PlotworldConfig(int roadSize, int plotSize, String biomeId,
                                  BlockConfig unclaimedBorderBlock, BlockConfig claimedBorderBlock,
                                  BlockConfig borderBlock, BlockConfig roadBlock, Layer[] layers) {

    }

    public record BlockConfig(String id, BlockStateConfig[] states) {

    }

    public record BlockStateConfig(String key, String value) {

    }

    public record Layer(int thickness, BlockConfig block) {

    }

    public record BlockPos(int x, int y, int z) {

    }

    public record ItemConfig(String itemId, int count, boolean hideAttributes, boolean glow, String headValue) {

    }
}


